packages/block-editor/src/components/list-view/block-select-button.js

```diff
@@ -87,7 +226,34 @@ function ListViewBlockSelectButton(
 					spacing={ 1 }
 				>
 					<span className="block-editor-list-view-block-select-button__title">
-						<Truncate ellipsizeMode="auto">{ blockTitle }</Truncate>
+						{ isEditingModeActive ? (
+							<>
+								<InputControl
+									ref={ inputRef }
+									value={ inputValue }
+									onChange={ ( nextValue ) => {
+										setInputValue( nextValue ?? '' );
+									} }
+									onBlur={ () => {
+										toggleLabelEditingMode( false );
+									} }
+									label={ __( 'Edit block name' ) }
+									hideLabelFromVision
+									aria-describedby={ inputDescriptionId }
+								/>
+								<VisuallyHidden>
+									<p id={ inputDescriptionId }>
+										{ __(
+											'Press the ENTER key to submit or the ESCAPE key to cancel.'
+										) }
+									</p>
+								</VisuallyHidden>
+							</>
```

Comment:
I don't think a component that's explicitly called 'Button' should be responsible for rendering an input.

Reply:
Yeh that's a fair point. It's pretty confusing. 

Hopefully we have the UX in place so now is a good time to do a refactor.

Reply:
We now have x2 dedicated components to handle edit mode and "button" mode.